esp-hal/src/aes/mod.rs

```diff
@@ -595,6 +751,480 @@ pub mod dma {
         }
     }
+
+    const AES_DMA_VTABLE: VTable<super::AesOperation> = VTable {
+        post: |driver, item| {
+            // Since we run the operation in `poll`, there is nothing to do here, besides accepting
+            // the work item. However, DS may be using the AES peripheral. We solve that
+            // problem (later) by taking a lock in `on_work_available`
+            let driver = unsafe { AesDmaBackend::from_raw(driver) };
+            if !driver.ensure_initialized() {
+                return false;
+            }
+            driver.start_processing(item);
+            true
+        },
+        poll: |driver, item| {
+            let driver = unsafe { AesDmaBackend::from_raw(driver) };
+            driver.poll_status(item)
+        },
+        cancel: |_driver, _item| {
+            // We can't abort an in-progress computation, or we may corrupt the block cipher's
+            // state.
+        },
+        stop: |driver| {
+            // Drop the AES driver to conserve power when there is nothig to do (or when the driver
+            // was stopped).
+            let driver = unsafe { AesDmaBackend::from_raw(driver) };
+            driver.deinitialize()
+        },
+    };
+
+    #[derive(Clone, Copy, PartialEq, Debug)]
+    #[cfg_attr(feature = "defmt", derive(defmt::Format))]
+    enum DmaState {
+        Idle,
+        WaitingForDma { processed_bytes: usize },
+    }
+
+    #[procmacros::doc_replace(
+        "dma_channel" => {
+            cfg(esp32s2) => "let dma_channel = peripherals.DMA_CRYPTO;",
+            _ => "let dma_channel = peripherals.DMA_CH0;"
+        }
+    )]
+    /// DMA-enabled AES processing backend.
+    ///
+    /// The backend will try its best to use hardware acceleration as much as possible. It will
+    /// fall back to CPU-based processing (equivalent to [`AesBackend`](super::AesBackend)) in some
+    /// cases, however:
+    ///
+    /// - When the block cipher is not implemented in hardware.
+    /// - When the data is not correctly aligned to the needs of the hardware (e.g. when using a
+    ///   stream cipher mode, the data length is not an integer multiple of 16 bytes).
+    ///
+    /// ## Example
+    ///
+    /// ```rust, no_run
+    /// # {before_snippet}
+    /// use esp_hal::aes::{AesContext, Operation, cipher_modes::Ecb, dma::AesDmaBackend};
+    /// #
+    /// # {dma_channel}
+    /// let mut aes = AesDmaBackend::new(peripherals.AES, dma_channel);
+    /// // Start the backend, which allows processing AES operations.
+    /// let _backend = aes.start();
+    ///
+    /// // Create a new context with a 128-bit key. The context will use the ECB block cipher mode.
+    /// // The key length must be supported by the hardware.
+    /// let mut ecb_encrypt = AesContext::new(
+    ///     Ecb,
+    ///     Operation::Encrypt,
+    ///     [
+    ///         b'S', b'U', b'p', b'4', b'S', b'e', b'C', b'p', b'@', b's', b'S', b'w', b'0', b'r',
+    ///         b'd', 0,
+    ///     ],
```

Comment:
```suggestion
    ///     *b"SUp4SeCp@sSw0rd\0",
```

Had to verify this suggestion hehe. https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=d450210091a2a3f292ff82afcb4a3f38